File: benchmarks/tsb/bench_date_offset_rollforward.ts
/**
 * Benchmark: DateOffset.rollforward / rollback / onOffset — snap dates to offset anchors.
 * Tests MonthEnd, BusinessDay, YearBegin, MonthBegin, YearEnd.
 * Outputs JSON: {"function": "date_offset_rollforward", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { MonthEnd, BusinessDay, YearBegin, MonthBegin, YearEnd } from "../../src/index.ts";

const SIZE = 5_000;
const WARMUP = 5;
const ITERATIONS = 50;

const monthEnd = new MonthEnd(1);
const bizDay = new BusinessDay(1);
const yearBegin = new YearBegin(1);
const monthBegin = new MonthBegin(1);
const yearEnd = new YearEnd(1);

const base = new Date(Date.UTC(2020, 0, 15));
const dates = Array.from({ length: SIZE }, (_, i) => new Date(base.getTime() + i * 86_400_000));

for (let i = 0; i < WARMUP; i++) {
  for (const d of dates.slice(0, 100)) {
    monthEnd.rollforward(d);
    monthEnd.rollback(d);
    monthEnd.onOffset(d);
    bizDay.rollforward(d);
    bizDay.rollback(d);
    bizDay.onOffset(d);
    yearBegin.rollforward(d);
    yearBegin.rollback(d);
    monthBegin.rollforward(d);
    monthBegin.rollback(d);
    yearEnd.rollforward(d);
    yearEnd.rollback(d);
  }
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  for (const d of dates) {
    monthEnd.rollforward(d);
    monthEnd.rollback(d);
    monthEnd.onOffset(d);
    bizDay.rollforward(d);
    bizDay.rollback(d);
    yearBegin.rollforward(d);
    yearBegin.rollback(d);
    monthBegin.rollforward(d);
    monthBegin.rollback(d);
    yearEnd.rollforward(d);
    yearEnd.rollback(d);
  }
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "date_offset_rollforward",
    mean_ms: Math.round((total / ITERATIONS) * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(total * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_date_offset_rollforward.py
"""Benchmark: DateOffset.rollforward / rollback / is_on_offset — snap dates to anchors.
Mirrors tsb bench_date_offset_rollforward.ts for pandas.tseries.offsets.
"""
import json, time
from datetime import datetime, timezone, timedelta
from pandas.tseries.offsets import MonthEnd, BusinessDay, YearBegin, MonthBegin, YearEnd

SIZE = 5_000
WARMUP = 5
ITERATIONS = 50

month_end = MonthEnd(1)
biz_day = BusinessDay(1)
year_begin = YearBegin(1)
month_begin = MonthBegin(1)
year_end = YearEnd(1)

import pandas as pd
base = pd.Timestamp("2020-01-15", tz="UTC")
dates = [base + timedelta(days=i) for i in range(SIZE)]

for _ in range(WARMUP):
    for d in dates[:100]:
        month_end.rollforward(d)
        month_end.rollback(d)
        month_end.is_on_offset(d)
        biz_day.rollforward(d)
        biz_day.rollback(d)
        biz_day.is_on_offset(d)
        year_begin.rollforward(d)
        year_begin.rollback(d)
        month_begin.rollforward(d)
        month_begin.rollback(d)
        year_end.rollforward(d)
        year_end.rollback(d)

start = time.perf_counter()
for _ in range(ITERATIONS):
    for d in dates:
        month_end.rollforward(d)
        month_end.rollback(d)
        month_end.is_on_offset(d)
        biz_day.rollforward(d)
        biz_day.rollback(d)
        year_begin.rollforward(d)
        year_begin.rollback(d)
        month_begin.rollforward(d)
        month_begin.rollback(d)
        year_end.rollforward(d)
        year_end.rollback(d)
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "date_offset_rollforward",
    "mean_ms": round(total / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total, 3),
}))
